Sources/GRPC/Serialization.swift

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2020, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import NIO
+import NIOFoundationCompat
+import SwiftProtobuf
+
+internal protocol MessageSerializer {
+  associatedtype Input
+
+  /// Serializes `input` into a `ByteBuffer` allocated using the provided `allocator`.
+  ///
+  /// The serialized buffer should have 5 leading bytes: the first must be zero, the following
+  /// four bytes are the `UInt32` encoded length of the serialized message. The bytes of the
+  /// serialized message follow.
```

Comment:
Woof, that's not a great API. Is there any reason not to instead update this to require a `serializeInto` method instead, and put the bytes we want there ourselves instead?

Reply:
I agree; it's a bit janky. I'd feel much worse about this if it were public API. The tradeoff with doing as you suggest is that you need to guess how large to allocate the buffer. 

Although if you just allocate with a minimum capacity of 5 bytes then it won't be _too_ bad since the copy with any required change in capacity will be small. Not sure about the tradeoff here.

Reply:
Or we could also break this up:

```swift
protocol MessageSerializer {
  associatedtype Input

  // returns the serialized size, or an estimate, if known.
  //
  // For `Message`: serialize and hold on to the `Data`, returns the exact size
  // For `GRPCPayload`: hold on to the input, return `nil`
  func willSerialize(_ input: Input) throws -> Int?
  
  // returns the actual number of bytes written.
  func writeSerializedBytes(into buffer: inout ByteBuffer) -> Int
}
```

Reply:
Yeah, I dunno. Perhaps it's better to leave it for now.